sway-lib-std/src/ufp64.sw

```diff
@@ -0,0 +1,247 @@
+use ::assert::assert;
+use ::math::*;
+use ::logging::*;
+use ::revert::revert;
+use ::u128::U128;
+
+pub struct UFP64 {
+    value: u64 
+}
+
+impl UFP64 {
+    pub fn denominator() -> u64 {
+        1 << 32
+    }
+
+    pub fn zero() -> UFP64 {
+        UFP64 {
+            value: 0,
+        }
+    }
+
+    /// The smallest value that can be represented by this type.
+    pub fn min() -> UFP64 {
+        UFP64 {
+            value: ~u64::min(),
+        }
+    }
+
+    /// The largest value that can be represented by this type,
+    pub fn max() -> UFP64 {
+        UFP64 {
+            value: ~u64::max(),
+        }
+    }
+
+    /// The size of this type in bits.
+    pub fn bits() -> u32 {
+        64
+    }
+}
+
+impl core::ops::Eq for UFP64 {
+    pub fn eq(self, other: Self) -> bool {
+        self.value == other.value
+    }
+}
+
+impl core::ops::Ord for UFP64 {
+    pub fn gt(self, other: Self) -> bool {
+        self.value > other.value
+    }
+
+    pub fn lt(self, other: Self) -> bool {
+        self.value < other.value
+    }
+}
+
+impl core::ops::Add for UFP64 {
+    /// Add a UFP64 to a UFP64. Panics on overflow.
+    pub fn add(self, other: Self) -> Self {
+        UFP64 {
+            value: self.value + other.value
+        }
+    }
+}
+
+impl core::ops::Subtract for UFP64 {
+    /// Subtract a UFP64 from a UFP64. Panics of overflow.
+    pub fn subtract(self, other: Self) -> Self {
+        // If trying to subtract a larger number, panic.
+        assert(!(self.value < other.value));
+
+        UFP64 {
+            value: self.value - other.value
+        }
+    }
+}
+
+impl core::ops::Multiply for UFP64 {
+    /// Multiply a UFP64 with a UFP64. Panics of overflow.
+    pub fn multiply(self, other: Self) -> Self {
+
+        let s_value = ~U128::from(0, self.value);
+        let o_value = ~U128::from(0, other.value);
+
+        let s_mul_o = s_value * o_value;
+        let res_u128 = s_mul_o >> 32;
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+
+        UFP64 {
+            value: res_u128.lower
+        }
+    }
+}
+
+impl core::ops::Divide for UFP64 {
+    /// Divide a UFP64 by a UFP64. Panics if divisor is zero.
+    pub fn divide(self, divisor: Self) -> Self {
+    let zero = ~UFP64::zero();
+        assert(divisor != zero);
+
+        let denominator = ~U128::from(0, ~Self::denominator());
+        let self_u128 = ~U128::from(0, self.value);
+        let divisor_u128 = ~U128::from(0, divisor.value);
+
+        let res_u128 = self_u128 * denominator / divisor_u128;
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        UFP64 {
+            value: res_u128.lower
+        }
+    }
+}
+
+
+
+impl UFP64 {
+
+    pub fn from_uint(uint: u64) -> UFP64 {
+        UFP64 {
+            value: ~Self::denominator() * uint,
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn recip(number: UFP64) -> Self {
```

Comment:
What is a "recip"? Is this the "reciprocal"? Name things properly.

Reply:
Yes. Naming was done according to that of Rust. So should I still rename?

Reply:
Well... I would personally say that Rust naming sucks here then and you should name things properly.
Wait for @adlerjohn to comment.

Reply:
We should generally follow Rust conventions and naming unless there's an explicit _functional_ reason to be different (e.g. reverting in the FuelVM is very different from panicking in Rust, and so should be named differently, and using `.unwrap()` is actually good practice in Sway). Purely _stylistic_ changes are subjective and should be avoided.

The word "sucks," well, sucks 😂. It doesn't fully capture the nuance: overly-verbose code can be worse than less-verbose code, and overly-terse code can be worse than more-verbose code. It's a continuum, not binary, and different people have different preferences.

Reply:
I will forever argue that abbreviations are bad in almost 100% of cases. The reader should have no knowledge of what they are looking at and be able to understand what is going on. Making them guess names is a sign of a terrible engineer.